{% extends "layout.html" %}

{% block nav_API %} <li class="nav-item active"> <a class="nav-link" href="{{ url_for('api_help') }}">API<span class="sr-only">(current)</span></a></li> {% endblock %}

{% block content %}


<div class="container-fluid">
<script src="{{ url_for('static', filename='js/prism.js') }}"></script>
<link rel="stylesheet" href="{{ url_for('static', filename='css/prism.css') }}">
<br>

<h1 class="row">Application Programming Interface</h1>
<div class="row">
    <div class="col-md-12">
        aGOtool has an application programming interface (API) which enables you to get the data without using the graphical user interface of the web page. The API is convenient if you need programmatic access. Please have a look at the examples.
    </div>
</div>
<br>

{#<h4>Entity types</h4>#}
{#    <br>#}
{#    <div class="col-md-5">#}
{#        {{content_entity_types | safe}}#}
{#    </div>#}
<div class="row">
    <div class="col-md-12">

<h2>Python 3 examples</h2>
    <pre><code class="language-python">
import requests
from io import StringIO
import pandas as pd
# call api_help for help and argument defaults
response = requests.get(r"https://agotool.org/api_help")
print(response.json())
url_ = r"https://agotool.org/api"
    </code></pre>

    <h4>Example #1 from STRING</h4>
<p><b>Protein name: trpA; Organism: Escherichia coli K12_MG1655</b></p>
        <pre><code class="language-python">
ENSPs = ['511145.b1260', '511145.b1261', '511145.b1262', '511145.b1263', '511145.b1264', '511145.b1812',
     '511145.b2551', '511145.b3117', '511145.b3360', '511145.b3772', '511145.b4388']
fg = "%0d".join(ENSPs)
result = requests.post(url_,
                   params={"output_format": "json",
                           "enrichment_method": "genome",
                           "taxid": 511145},
                   data={"foreground": fg})
result_json = result.json()
print("Enrichment results for the following entity types was calculated: \n", result_json.keys())
            </code></pre><br>

    <h4>Example #2 from STRING</h4>
    Use iPython or Jupyter notebook or Jupyter lab to interactively view results as a pandas data frame.<br>
        <p><b>Protein name: CDC15; Organism: Saccharomyces cerevisiae</b></p>
        <pre><code class="language-python">
ENSPs = ['4932.YAR019C', '4932.YFR028C', '4932.YGR092W', '4932.YHR152W', '4932.YIL106W', '4932.YJL076W',
     '4932.YLR079W', '4932.YML064C', '4932.YMR055C', '4932.YOR373W', '4932.YPR119W']
fg = "%0d".join(ENSPs)
result = requests.post(url_,
                   params={"output_format": "tsv",
                           "enrichment_method": "genome",
                           "taxid": 4932},
                   data={"foreground": fg})
result = result.text
df = pd.read_csv(StringIO(result), sep='\t')
df.head()
        </code></pre><br>
    <h4>Example #3 from STRING</h4>
<p><b>Protein name: trpA; Organism: Escherichia coli K12_MG1655</b></p>
        <pre><code class="language-python">
ENSPs = ['10090.ENSMUSP00000001812', '10090.ENSMUSP00000002708', '10090.ENSMUSP00000021921',
     '10090.ENSMUSP00000025791', '10090.ENSMUSP00000026474', '10090.ENSMUSP00000030443',
     '10090.ENSMUSP00000054837', '10090.ENSMUSP00000084430', '10090.ENSMUSP00000099623',
     '10090.ENSMUSP00000106137', '10090.ENSMUSP00000107498']
fg = "%0d".join(ENSPs)
result = requests.post(url_,
                   params={"output_format": "json",
                           "enrichment_method": "compare_samples",
                           "limit_2_entity_type": "-21;-51;-52", # comma separate desired entity_types (e.g. GO biological process; UniProt keyword; SMART) , default get all available.
                           "taxid": 10090},
                   data={"foreground": fg,
                         "background": fg})
# since the foreground equals the background completely no significant enrichment will result from this analysis
# it is meant to show how to show how to use the method call
result_json = result.json()
import ast
ast.literal_eval(result_json["-51"])
          </code></pre><br>

    <h4>Example #4</h4>
<p><b>Characterize foreground, get functional information </b></p>
        <pre><code class="language-python">
ENSPs = ["9606.ENSP00000266970", "9606.ENSP00000278616"]
fg = "%0d".join(ENSPs)
result = requests.post(url_,
                   params={"output_format": "tsv",
                           "filter_parents": False,
                           "enrichment_method": "characterize_foreground"},
                   data={"foreground": fg})
df = pd.read_csv(StringIO(result.text), sep='\t')
df.head()
          </code></pre><br>

</div>
</div>
</div>
<br><br>

<h1 class="row"> API parameters</h1>
    <div class="row">
        <div class="col-md-12">
            <ul>
                <li class="bulletpoints_li"><b><em>taxid</em></b>:<a href="https://www.ncbi.nlm.nih.gov/taxonomy"> NCBI taxon identifier</a> (e.g. "9606" for Homo sapiens), default=None (not set).
                <li class="bulletpoints_li"><b><em>species</em></b>: deprecated, please use "taxid" instead.
                <li class="bulletpoints_li"><b><em>organism</em></b>: deprecated, please use "taxid" instead.
                <li class="bulletpoints_li"><b><em>output_format</em></b>: The desired format of the output, one of {tsv, tsv-no-header, json, xml}, default="tvs".
                <li class="bulletpoints_li"><b><em>filter_parents</em></b>: Remove parent terms (keep GO terms and UniProt Keywords of lowest leaf) if they are associated with exactly the same foreground. For a more detailed description see Parameters --> Report options --> Filter redundant parent terms. Default=True.
                <li class="bulletpoints_li"><b><em>filter_foreground_count_one</em></b>: Keep only those terms with foreground_count > 1. Default=True.
                <li class="bulletpoints_li"><b><em>filter_PMID_top_n</em></b>: Filter the top n PMIDs (e.g. 100, default=100), sorting by low p value and recent publication date. Default=100.
                <li class="bulletpoints_li"><b><em>caller_identity</em></b>: Your identifier for us e.g. "www.my_awesome_app.com". Default=None (this parameter exists to comply with STRING).
                <li class="bulletpoints_li"><b><em>FDR_cutoff</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>
                <li class="bulletpoints_li"><b><em>taxid</em></b>

{#parser.add_argument("FDR_cutoff", type=float, help="False Discovery Rate cutoff (cutoff for multiple testing corrected p values) e.g. 0.05, default=0.05 meaning 5%. Set to 1 for no cutoff.", default=0.05)#}
{#parser.add_argument("limit_2_entity_type", type=str, help="Limit the enrichment analysis to a specific or multiple entity types, e.g. '-21' (for GO molecular function) or '-21;-22;-23;-51' (for all GO terms as well as UniProt Keywords).", default="-20;-21;-22;-23;-51;-52;-54;-55;-56;-57;-58") # -53 missing for UniProt version#}
{#parser.add_argument("privileged", type=str, default="False")#}
{#parser.add_argument("foreground", type=str, help="ENSP identifiers for all proteins in the test group (the foreground, the sample, the group you want to examine for GO term enrichment) "#}
{#         "separate the list of Accession Number using '%0d' e.g. '4932.YAR019C%0d4932.YFR028C%0d4932.YGR092W'",#}
{#    default=None)#}
{#parser.add_argument("background", type=str,#}
{#    help="ENSP identifiers for all proteins in the background (the population, the group you want to compare your foreground to) "#}
{#         "separate the list of Accession Number using '%0d'e.g. '4932.YAR019C%0d4932.YFR028C%0d4932.YGR092W'",#}
{#    default=None)#}
{#parser.add_argument("background_intensity", type=str,#}
{#    help="Protein abundance (intensity) for all proteins (copy number, iBAQ, or any other measure of abundance). "#}
{#         "Separate the list using '%0d'. The number of items should correspond to the number of Accession Numbers of the 'background'"#}
{#         "e.g. '12.3%0d3.4' ",#}
{#    default=None)#}
{#parser.add_argument("enrichment_method", type=str,#}
{#    help="""'genome': provided foreground vs genome;#}
{#    'compare_samples': Foreground vs Background (no abundance correction); #}
{#    'characterize_foreground': list all functional annotations for provided foreground;#}
{#    'abundance_correction': Foreground vs Background abundance corrected""",#}
{#    default="characterize_foreground")#}
{#parser.add_argument("goslim", type=str, help="GO basic or a slim subset {generic, agr, aspergillus, candida, chembl, flybase_ribbon, metagenomics, mouse, pir, plant, pombe, synapse, yeast}. Choose between the full Gene Ontology ('basic') or one of the GO slim subsets (e.g. 'generic' or 'mouse'). GO slim is a subset of GO terms that are less fine grained. 'basic' does not exclude anything, select 'generic' for a subset of broad GO terms, the other subsets are tailor made for a specific taxons / analysis (see http://geneontology.org/docs/go-subset-guide)", default="basic")#}
{#parser.add_argument("o_or_u_or_both", type=str, help="over- or under-represented or both, one of {overrepresented, underrepresented, both}. Choose to only test and report overrepresented or underrepresented GO-terms, or to report both of them.", default="overrepresented")#}
{#parser.add_argument("num_bins", type=int, help="The number of bins created based on the abundance values provided. Only relevant if 'Abundance correction' is selected.", default=100)#}
{#parser.add_argument("fold_enrichment_for2background", type=float, help="Apply a filter for the minimum cutoff value of fold enrichment foreground/background.",default=0)#}
{#parser.add_argument("p_value_cutoff", type=float, help="Apply a filter (value between 0 and 1) for maximum cutoff value of the uncorrected p value. '1' means nothing will be filtered, '0.01' means all uncorected p_values <= 0.01 will be removed from the results (but still tested for multiple correction).", default=1)#}
{#parser.add_argument("multiple_testing_per_etype", type=bool, help="If True calculate multiple testing correction separately per entity type (functional category), in contrast to performing the correction together for all results.", default=True)#}
{#parser.add_argument("score_cutoff", type=float, help="Apply a filter for the minimum cutoff value of the textmining score. This cutoff is only applied to the 'characterize_foreground' method, and does not affect p values. Default = 3.", default=3)#}
{#parser.add_argument("foreground_replicates", type=int, help="'foreground_replicates' is an integer, defines the number of samples (replicates) of the foreground.", default=10)#}
{#parser.add_argument("background_replicates", type=int, help="'background_replicates' is an integer, defines the number of samples (replicates) of the background.", default=10)    #}


        </div>
    </div>
</div>

<br><br><br><br>


{% endblock content %}